File: ctci/tmp/chapter_3_stacks_and_queues/python/stack.py
# First in, last out.
class Stack:
    def __init__(self):
        """
        Stack constructor.
        """
        self.stack = []
        self.length = 0

    def push(self, value: int):
        """
        Pushes an element to the top of the stack.
        """
        self.stack.append(value)
        self.length += 1

    def pop(self) -> int:
        """
        Pops an element off the top of the stack.
        """
        self.length -= 1
        return self.stack.pop()

    def peek(self) -> int:
        """
        Peeks at the element at the top of the stack.
        """
        return self.stack[self.length - 1]

    def is_empty(self) -> bool:
        """
        Determines if a stack is empty.
        """
        return self.length == 0


if __name__ == "main":
    s = Stack()
    for i in [1, 2, 3, 4, 5]:
        s.push(i)
    print(s.stack)  # [1, 2, 3, 4, 5]
    print(s.pop())  # 5
    print(s.peek())  # 4
    print(s.stack)  # [1, 2, 3, 4]

File: ctci/tmp/chapter_3_stacks_and_queues/js/stack.js
/* First in, last out. */
class Stack {
  constructor() {
    /**
     * Stack constructor.
     */
    this.stack = []
    this.length = 0
  }

  /**
   * Pushes an element to the top of the stack.
   * @param   {number}  value
   */
  push(value) {
    this.stack.push(value)
    this.length++
  }

  /**
   * Pops an element off the top of the stack.
   * @returns {number}
   */
  pop() {
    this.length--
    return this.stack.pop()
  }

  /**
   * Determines if a stack is empty.
   * @returns {boolean}
   */
  isEmpty() {
    return this.length === 0
  }

  /**
   * Peeks at the element at the top of the stack.
   * @returns {number}
   */
  peek() {
    return this.stack[this.length - 1]
  }
}

module.exports = {
  Stack
}

const st = new Stack()
console.log(st) // Stack { stack: [], length: 0 }
st.push(1)
st.push(2)
st.push(4)
console.log(st.peek()) // 4
st.pop()
console.log(st) // Stack { stack: [ 1, 2 ], length: 2 }
